### PR TITLE
Run mypy type checking on the codebase with Travis

### PR DESCRIPTION
**Pull request by jenshnielsen**

### CODE REVIEW

File: qcodes/data/data_set.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ class DataSet(DelegateAttributes):
     default_formatter = GNUPlotFormat()
     location_provider = FormatLocation()
 
-    background_functions = OrderedDict()
+    background_functions = OrderedDict()  # type: OrderedDict
```
Comment (jenshnielsen) [inline qcodes/data/data_set.py]: This should probably be `OrderedDict[Callable]`

File: qcodes/instrument/parameter.py [hunk 2]
```diff
@@ -251,7 +253,7 @@ def __call__(self, *args, **kwargs):
                                           ' Parameter {}'.format(self.name))
 
     def snapshot_base(self, update: bool=False,
-                      params_to_skip_update: Sequence[str]=None) -> dict:
+                      params_to_skip_update: Sequence[str]=None) -> Dict[str, Any]:
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: I guess this could be Dict[str, Parameter]

File: qcodes/plots/pyqtgraph.py [hunk 3]
```diff
@@ -49,7 +50,8 @@ class QtPlot(BasePlot):
     # close event on win but this is difficult with remote proxy process
     # as the list of plots lives in the main process and the plot locally
     # in a remote process
-    plots = deque(maxlen=qcodes.config['gui']['pyqtmaxplots'])
+    max_len = qcodes.config['gui']['pyqtmaxplots'] # type: int
+    plots = deque(maxlen=max_len) # type: Deque['QtPlot']
```
Comment (jenshnielsen) [inline qcodes/plots/pyqtgraph.py]: This could probably also be a remote proxy?

**Comment by jenshnielsen** [comment]
This is still using python 3.5 type variable annotation( with comments) once we drop that it should be converted to 3.6 style

File: qcodes/instrument/parameter.py (hunk 2 above)
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: Or probably not. It's more complicated

**Comment by Dominik-Vogel** [comment]
great job!

**Comment by jenshnielsen** [comment]
@Dominik-Vogel not so great at the moment :( build failing

**Pull request merged**